File: 08-tries/00211-Design_Add_And_Search_Words_Data_Structure/00211-design_add_and_search_words_data_structure.cpp
// 211: Design Add And Search Words Data Structure
// https://leetcode.com/problems/design-add-and-search-words-data-structure/


#include <iostream>
#include <vector>
#include <string>

using namespace std;


class WordDictionary {
private:
    vector<WordDictionary*> children;
    bool isEndOfWord;

public:
    WordDictionary(): isEndOfWord(false){
        this->children = vector<WordDictionary*>(26, nullptr);
        this->isEndOfWord = false;
    }
    

    void addWord(string word) {
        WordDictionary *current = this;

        for (char c: word) {
            if (current->children[c - 'a'] == nullptr)
                current->children[c - 'a'] = new WordDictionary();
            current = current->children[c - 'a'];
        }

        current->isEndOfWord = true;
    }


    bool search(string word) {
        WordDictionary *current = this;

        for (int i=0; i<word.length(); i++){
            char c = word[i];

            if (c == '.') {
                for (auto ch: current->children)
                    if (ch && ch->search(word.substr(i+1))) return true;
                return false;
            }

            if (current->children[c - 'a'] == nullptr) return false;
            current = current->children[c - 'a'];
        }

        return current && current->isEndOfWord;
    }
};


int main() {
    WordDictionary* wd = new WordDictionary();

    // OPERATIONS
    wd->addWord("bad");
    wd->addWord("dad");
    wd->addWord("mad");
    cout << (wd->search("pad") ? "true" : "false") << endl;
    cout << (wd->search("bad") ? "true" : "false") << endl;
    cout << (wd->search(".ad") ? "true" : "false") << endl;
    cout << (wd->search("b..") ? "true" : "false") << endl;

    return 0;
}

File: 08-tries/00211-Design_Add_And_Search_Words_Data_Structure/00211-design_add_and_search_words_data_structure.java
// 211: Design Add And Search Words Data Structure
// https://leetcode.com/problems/design-add-and-search-words-data-structure/



class WordDictionary {
    private WordDictionary[] children;
    private boolean isEndOfWord;

    public WordDictionary() {
        this.children = new WordDictionary[26];
        this.isEndOfWord = false;
    }
    

    public void addWord(String word) {
        WordDictionary current = this;

        for (char c: word.toCharArray()) {
            if (current.children[c - 'a'] == null)
                current.children[c - 'a'] = new WordDictionary();
            current = current.children[c - 'a'];
        }

        current.isEndOfWord = true;
    }
    

    public boolean search(String word) {
        WordDictionary current = this;

        for (int i=0; i<word.length(); i++){
            char c = word.charAt(i);

            if (c == '.') {
                for (WordDictionary ch: current.children)
                    if( ch != null && ch.search(word.substring(i+1))) return true;
                return false;
            }

            if (current.children[c - 'a'] == null) return false;
            current = current.children[c - 'a'];
        }

        return current != null && current.isEndOfWord;
    }

    public static void main(String[] args) {
        WordDictionary wd = new WordDictionary();

        // OPERATIONS
        wd.addWord("bad");
        wd.addWord("dad");
        wd.addWord("mad");
        System.out.println((wd.search("pad") ? "true" : "false"));
        System.out.println((wd.search("bad") ? "true" : "false"));
        System.out.println((wd.search(".ad") ? "true" : "false"));
        System.out.println((wd.search("b..") ? "true" : "false"));
    }
}
